controller/src/main/java/io/pravega/controller/server/SegmentStoreConnectionManager.java

```diff
@@ -0,0 +1,653 @@
+                            .removalListener((RemovalListener<PravegaNodeUri, SegmentStoreConnectionPool>) removalNotification -> {
+                                // Whenever a connection manager is evicted from the cache call shutdown on it. 
+                                removalNotification.getValue().shutdown();
+                            })
+                            .build(new CacheLoader<PravegaNodeUri, SegmentStoreConnectionPool>() {
+                                @Override
+                                @ParametersAreNonnullByDefault
+                                public SegmentStoreConnectionPool load(PravegaNodeUri nodeUri) {
+                                    return new SegmentStoreConnectionPool(nodeUri, clientCF);
+                                }
+                            });
+
+    }
+
+    public SegmentStoreConnectionPool getPool(PravegaNodeUri uri) {
+        return cache.getUnchecked(uri);
+    }
+
+    /**
+     * This is a connection manager class to manage connection to a given segmentStore node identified by PravegaNodeUri. 
+     * It maintains a pool of available connections and creates a maximum number of concurrent connections. 
+     * Users can request for connection from this class and it will opportunistically use existing connections or create new 
+     * connections to a given segment store server and return the connection.
+     * It ensures that there are only a limited number of concurrent connections created. If more users request for connection
+     * it would add them to wait queue and as connections become available (when existing connections are returned),  
+     * this class opportunistically tries to reuse the returned connection to fulfil waiting requests. If there are no 
+     * waiting requests, this class tries to maintain an available connection pool of predetermined size. If more number of 
+     * connections than max available size is returned to it, the classes closes those connections to free up resources.  
+     *
+     * It is important to note that this is not a connection pooling class. 
+     * It simply reuses already created connections to send additional commands over it. It doesnt multiplex commands on the 
+     * same connection concurrently.  
+     * As users finish their processing, they should return the connection back to this class. 
+     *
+     * The connectionManager can be shutdown as well. However, the shutdown trigger does not prevent from callers to attempt 
+     * to create new connections and new connections will be served. Shutdown ensures that 
+     */
+    static class SegmentStoreConnectionPool {
+        private static final int MAX_CONCURRENT_CONNECTIONS = 500;
+        private static final int MAX_AVAILABLE_CONNECTIONS = 100;
+        private final PravegaNodeUri uri;
+        private final Object lock = new Object();
+        @GuardedBy("lock")
+        private final ArrayDeque<ConnectionObject> availableConnections;
```

Comment:
This class seems very elaborate given that it's not doing true pooling which is going to be done by another component underneath it. Could we simplify things by instead of having a pool having a 'reusableConnection' which we just return to the cache directly when done?

Reply:
yes, this class is slightly more complicated. 
It ensures 2 things -
1. limit the max concurrent connections created to a given segment store instance
2. once connections are returned to the pool, keep a ceiling on number of available connections. 

but i had a discussion with @shrids and with the multiplexing of requests over single connection, i dont need to maintain any connection logic. 
the connection factory should automatically make use of the same connection to send multiple requests. 
So once #3531 is merged, it will simplify things quite a bit and this class can be discarded. 
So let me wait on that change and once its merged, i will update this PR.

Reply:
On @andreipaduroiu 's suggestion i have moved the resource pooling logic to a new ResourcePool class.